crates/l2/monitor/app.rs

```diff
@@ -53,9 +58,126 @@ pub struct EthrexMonitor {
     pub store: Store,
     pub rollup_store: StoreRollup,
     pub last_scroll: Instant,
+    pub last_tick: Instant,
+}
+
+#[derive(Clone)]
+pub struct EthrexMonitorState {
+    widget: EthrexMonitorWidget,
+    terminal: Arc<Mutex<Terminal<CrosstermBackend<io::Stdout>>>>,
+    cancelation_token: CancellationToken,
+}
+
+pub struct EthrexMonitor {}
+
+#[derive(Clone, Debug)]
+pub enum CastInMessage {
+    Monitor,
+}
+
+#[derive(Clone, Debug)]
+pub enum CallInMessage {
+    Finished,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum OutMessage {
+    Done,
+    ShouldQuit(bool),
 }
 
 impl EthrexMonitor {
+    pub async fn spawn(
+        sequencer_state: SequencerState,
+        store: Store,
+        rollup_store: StoreRollup,
+        cfg: &SequencerConfig,
+        cancelation_token: CancellationToken,
+    ) -> Result<(), MonitorError> {
+        let widget = EthrexMonitorWidget::new(sequencer_state, store, rollup_store, cfg).await?;
+        let state = EthrexMonitorState {
+            widget,
+            terminal: Arc::new(Mutex::new(setup_terminal()?)),
+            cancelation_token,
+        };
+        let mut ethrex_monitor = EthrexMonitor::start_blocking(state);
+        ethrex_monitor
+            .cast(CastInMessage::Monitor)
+            .await
+            .map_err(MonitorError::GenServerError)?;
+        Ok(())
+    }
+
+    pub async fn monitor(&self, state: &mut EthrexMonitorState) -> Result<(), MonitorError> {
+        let mut terminal = state.terminal.lock().await;
+        let widget = &mut state.widget;
+
+        widget.draw(&mut terminal)?;
+
+        let timeout =
+            Duration::from_millis(widget.tick_rate).saturating_sub(widget.last_tick.elapsed());
+        if !event::poll(timeout)? {
+            widget.on_tick().await?;
+            widget.last_tick = Instant::now();
+            return Ok(());
+        }
+        let event = event::read()?;
+        if let Some(key) = event.as_key_press_event() {
+            widget.on_key_event(key.code);
+        }
+        if let Some(mouse) = event.as_mouse_event() {
+            widget.on_mouse_event(mouse.kind);
+        }
+
+        Ok(())
+    }
+
+    pub async fn terminate(&self, state: &EthrexMonitorState) {
```

Comment:
Sorry, I commented the wrong name on this function, it's called `teardown()` and it is part of the [Trait](https://github.com/lambdaclass/spawned/blob/f4ddd843501165bf8aefef004ffb49ac04b2ef48/concurrency/src/tasks/gen_server.rs#L193), so you don't need to call it explicitly